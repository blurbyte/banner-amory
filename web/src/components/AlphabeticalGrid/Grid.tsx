import BaseGrid from '../Grid';

const Grid = BaseGrid.withComponent('div');

export default Grid;
