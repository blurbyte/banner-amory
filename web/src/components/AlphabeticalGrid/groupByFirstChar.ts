// group items by their name first character

import { ItemBasic } from '@sharedTypes/Item';

export type GroupedItems = {
  [char: string]: ItemBasic[];
};

function groupByFirstChar(items: ItemBasic[]) {
  return items.reduce(
    (groupedItems, item) => {
      const key = item.slug.charAt(0);
      groupedItems[key] = groupedItems[key] || [];
      groupedItems[key].push(item);
      return groupedItems;
    },
    {} as GroupedItems
  );
}

export default groupByFirstChar;
