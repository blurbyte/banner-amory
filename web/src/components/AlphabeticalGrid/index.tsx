// items grid ordered alphabetically

import * as React from 'react';

import { ItemBasic } from '@sharedTypes/Item';
import groupByFirstChar from './groupByFirstChar';
import Grid from './Grid';
import Group from '../AlphabeticalGridGroup';

type AlphabeticalGridProps = {
  items: ItemBasic[];
};

class AlphabeticalGrid extends React.Component<AlphabeticalGridProps> {
  render() {
    const { items } = this.props;
    const groupedItems = groupByFirstChar(items);
    const characters = Object.keys(groupedItems).sort();

    return (
      <Grid>
        {characters.map(char => {
          return <Group key={`${char}-items`} char={char} items={groupedItems[char]} />;
        })}
      </Grid>
    );
  }
}

export default AlphabeticalGrid;
