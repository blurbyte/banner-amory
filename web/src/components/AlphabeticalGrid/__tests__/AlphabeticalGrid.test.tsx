import * as React from 'react';
import * as renderer from 'react-test-renderer';

import AlphabeticalGrid from '../index';
import { items } from './mocks';

test('renders items grouped by first name character', () => {
  const component = renderer.create(<AlphabeticalGrid items={items} />);
  expect(component.toJSON()).toMatchSnapshot();
});
