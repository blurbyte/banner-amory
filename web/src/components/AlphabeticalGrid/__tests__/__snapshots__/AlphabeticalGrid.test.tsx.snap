// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders items grouped by first name character 1`] = `
<div
  className="sc-bdVaJa iDpojy"
>
  <div
    className="sc-jTzLTM gKdRlC"
  >
    <p
      className="sc-cSHVUG fTOKnP sc-jzJRlG dGTYFu"
    >
      t
    </p>
    <ul
      className="sc-kAzzGY kHKYvg"
    >
      <li
        className="sc-ifAKCX eAYGcP"
      >
        <a
          className="sc-EHOje jQSOuj"
          href="/item/taylors-bracelet"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="sc-bwzfXH elpSZb"
          >
            <img
              className="sc-htpNat kvsjSp"
              src="undefinedtaylors-bracelet.png"
            />
          </div>
          <p
            className="sc-gqjmRU hyXOOH sc-gzVnrw dJHat"
          >
            1
          </p>
        </a>
      </li>
      <li
        className="sc-ifAKCX eAYGcP"
      >
        <a
          className="sc-EHOje jQSOuj"
          href="/item/thunder-ball"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="sc-bwzfXH elpSZb"
          >
            <img
              className="sc-htpNat kvsjSp"
              src="undefinedthunder-ball.png"
            />
          </div>
          <p
            className="sc-gqjmRU hyXOOH sc-gzVnrw dJHat"
          >
            2
          </p>
        </a>
      </li>
    </ul>
  </div>
  <div
    className="sc-jTzLTM ldRjeR"
  >
    <p
      className="sc-cSHVUG fTOKnP sc-jzJRlG dGTYFu"
    >
      v
    </p>
    <ul
      className="sc-kAzzGY cMQPDM"
    >
      <li
        className="sc-ifAKCX eAYGcP"
      >
        <a
          className="sc-EHOje jQSOuj"
          href="/item/vibrant-touch"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="sc-bwzfXH elpSZb"
          >
            <img
              className="sc-htpNat kvsjSp"
              src="undefinedvibrant-touch.png"
            />
          </div>
          <p
            className="sc-gqjmRU hyXOOH sc-gzVnrw dJHat"
          >
            3
          </p>
        </a>
      </li>
    </ul>
  </div>
</div>
`;
